### ISSUE TITLE
Prompting for encryption password on unencrypted database

### ISSUE THREAD

**Issue by joelworsham** [sqlitebrowser/sqlitebrowser#932]
### Details for the issue

I download an sql file from a site backup on WP Engine. Multiple agents ensure me the file is NOT encrypted. One specific response he had when asked was:

> Positive because when I cat the file out its not encrypted at all.
> It’s just plaintext SQL

I download the file and try to open it and it prompts me for the password.

~[](http://d.pr/i/89yz+)

I click "ok" and it reloads that same prompt.

I click "cancel" and it tells me the file is invalid:

~[](http://d.pr/i/Shga+)

#### I'm opening this issue because:

- [ ] DB4S is crashing
- [X ] DB4S has a bug
- [ ] DB4S needs a feature
- [ ] DB4S has another problem

#### I'm using DB4S on:

- [ ] Windows: ( _version:_ ___ )
- [ ] Linux: ( _distro:_ ___ )
- [X] Mac OS: ( _version:10.11.6)
- [ ] Other: ___

#### I'm using DB4S version:

- [X] 3.9.1
- [ ] 3.9.0
- [ ] Other: ___

#### I have also:

- [ ] Tried out the nightly version: https://github.com/sqlitebrowser/sqlitebrowser#windows
- [X] Searched for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by revolter** [comment]
I think you're trying to open up an `.sql` file, and not a database `.sqlite` file.

**Comment by revolter** [comment]
I wonder if we can detect if the file is a database or not, and print something like "Not a database".

**Comment by chrisjlocke** [comment]
Just to elaborate this answer; a database consists of tables, fields and records, etc.  A .sql file is instructions (in SQL) to recreate that database.  As such, DB4S cannot 'open' that .sql file, but it can import it, so you can create a blank database, then import that .sql file.  :)

**Comment by justinclift** [comment]
Ahhh, yeah.  That definitely sounds like a text file with SQL statements in it, not a SQLite database file itself.

@joelworsham Not to worry though, you should be able to import it using **File** → **Import** → **Database from SQL file...**

![screen shot 2017-01-13 at 15 44 38](https://cloud.githubusercontent.com/assets/406299/21935937/4db468a4-d9a7-11e6-9dae-5a0974838417.png)

Are you ok to try that out and let us know if it worked for you? :smile:

**Comment by chrisjlocke** [comment]
> I wonder if we can detect if the file is a database or not

Should be easy to detect if the first 100 characters or so of the file contains pure valid SQL commands .... usually only 'drop table if not exists' and all that jazz...

**Comment by justinclift** [comment]
@revolter *I wonder if we can detect if the file is a database or not, and print something like "Not a database".*

Yeah, we kind of do already.  We first try to detect if it's a database file, and if it's not recognisable then we assume it's an encrypted file instead.  So, we then pop open a dialog asking for a password (etc).

I suppose we could add a step in between those, where we try to detect if the file is standard text in SQL statement format, and automatically open the Import dialog if it is.  No idea how practical that would be to implement though... :wink:

**Comment by joelworsham** [comment]
Thanks guys. I thought it was a valid file, but it was indeed just a dump, plaintext. It would be nice if the program could give a better response to this though. I'm willing to bet many others have done exactly what I did. Perhaps you can somehow tell that it is a DB dump and tell the user something like:

> It appears you're trying to import a database dump. Instead, import the file blah blah

**Comment by nico-ha** [comment]
Hi everybody, I just had this dialog appear on a corrupted but unencrypted database file (sqlite shell says "Error: database disk image is malformed"). So it seems like DB4S just asks for a password as soon as it is unable to read a database file, no matter the actual reason.

**Comment by justinclift** [comment]
@nico-ha Yep, that's definitely the behaviour at the moment.  We'll likely put some effort into improving that, so it tries a few other file types (csv, sql, cert) before jumping to the conclusion it's encrypted.

Not sure when, but we'll get around to it. :smile:

**Comment by nico-ha** [comment]
@justinclift Thanks for the explanation! No big deal, just good to know 👍

**Comment by mgrojo** [comment]
Just for the record. This just happened to me after a crash of DB4S. It wasn't the first that I had a crash with this database, but this was the first time that the password dialog opened after opening the DB file again in a new execution. I noticed that a journal file was in the same directory as the database file. I moved it to other location and now I can open again the database.

**Comment by buckleykm** [comment]
> I noticed that a journal file was in the same directory as the database file. I moved it to other location and now I can open again the database.

This was the fix for me!

**Comment by justinclift** [comment]
Ok.

1. Backup both files (just in case).  Really do this.  Not joking. :wink:
2. Move or rename the journal file, so it doesn't have the same name as the database.
3. Restart DB Browser for SQLite, then try opening the database again.

In theory (!), when the journal file isn't there, then the main database file should be opened ok.

But it *might* be missing any of the most recent changes.  And if the database file itself is damaged, this won't work either.

It's worth trying though. :smile:

**Comment by azeemrh** [comment]
That worked fine.

Thanks alot, Justin! :)

**Comment by justinclift** [comment]
You're welcome @azeemrh. :smile:

**Comment by marius-stanescu** [comment]
> Ok.
> 
> 1. Backup both files (just in case).  Really do this.  Not joking. 😉
> 2. Move or rename the journal file, so it doesn't have the same name as the database.
> 3. Restart DB Browser for SQLite, then try opening the database again.
> 
> In theory (!), when the journal file isn't there, then the main database file should be opened ok.
> 
> But it _might_ be missing any of the most recent changes. And if the database file itself is damaged, this won't work either.
> 
> It's worth trying though. 😄

But removing the journal file might leave the database in a corrupt state! (https://www.sqlite.org/tempfiles.html#rollback_journals)

**Comment by justinclift** [comment]
Hmmm, wasn't that clearly enough pointed out?

**Comment by chrisjlocke** [comment]
> Removing the journal file might leave the database in a corrupt state

If DB4S is asking for a password, then its already in a corrupt state.  There are two forms of 'corrupted'.  There is 'mangled database file' where the file itself if not a valid SQLite file, and 'mangled data' where the database is fine and working, but the data is not complete, eg, a user with a 'department_id' of 65, where there is no longer a department_id of 65 in the department table ... as it was in the journal file.
Deleting the journal file usually does the latter damage, not the former damage.

**Comment by tokenflow** [comment]
Same problem here

<img width="731" alt="Screen Shot 2019-05-13 at 6 17 36 PM" src="https://user-images.githubusercontent.com/8538953/57628694-7b4cd100-75ab-11e9-8ee9-688479c6b5a6.png">

**Comment by justinclift** [comment]
Excellent. :smile:

**Comment by tenuki** [comment]
I had the same problem. In my case it was the file/database (.sqlite3), was created with a feature not available in the browser version I had (generated columns was in my case).  The newly installed version didn't overwrite the version I had (in user's APPDATA).. and that was the program version which auto open...

**Comment by chrisjlocke** [comment]
Thanks for the helpful information.  Generated columns were introduced in SQLite 3.31 (Jan 2020), so if you were using a version of SQLCipher before then, then that would make sense.

**Comment by AshokPMurthy** [comment]
COPY the file (mydatabase.db3) to your "Downloads" folder. 
Now open the file with "DB Browser for SQLite".
It should open without a problem. Good Luck!
